addons/frag/functions/fnc_dev_debugAmmo.sqf

```diff
@@ -15,67 +19,86 @@
  *
  * Public: No
  */
-
 params [
-    ["_debugMissing", true, [false]],
-    ["_debugForce", false, [false]],
-    ["_debugSkippedFragPower", 30, [0]]
+    ["_logAll", false, [false]],
+    ["_onlyShotAmmoTypes", false, [false]],
+    ["_csvFormat", false, [false]]
 ];
 
 diag_log text format ["~~~~~~~~~~~~~Start [%1]~~~~~~~~~~~~~", _this];
+if (_csvFormat) then {
+    diag_log text format ["ammo,gurney_c,gurney_m,gurney_k,gurney_gC,fragTypes,fragCount,Inheritance"];
+};
 
-private _allMagsConfigs = configProperties [configFile >> "CfgMagazines", "isClass _x", true];
-private _processedCfgAmmos = [];
-
-{
-    private _ammo = toLower getText (_x >> "ammo");
-    if (_ammo != "" && {!(_ammo in _processedCfgAmmos)}) then {
-        _processedCfgAmmos pushBack _ammo;
-
-        //Ignore mines/bombs
-        if (_ammo isKindOf "TimeBombCore") exitWith {};
+// Gather all configs, either those that could be created from firing or all classes
+private _allAmmoConfigs = [];
+if (_onlyShotAmmoTypes) then {
+    private _configSearchFunc = {
+        params [
+            ["_ammo", "", [""]]
+        ];
+        if (_ammo isEqualTo "" || {_ammo in _allAmmoConfigs}) exitWith {};
+        _allAmmoConfigs pushBack _ammo;
+        private _ammoConfig = configFile >> "cfgAmmo" >> _ammo;
+        private _subMunit = toLowerANSI getText (_ammoConfig >> "submunitionAmmo");
+        if (_subMunit isNotEqualTo "") then {
+            _subMunit = getArray (_ammoConfig >> "submunitionAmmo");
+            for "_i" from 0 to count _subMunit - 1 do {
+                if (_i mod 2 == 0) then {
+                    [toLowerANSI (_subMunit#_i)] call _configSearchFunc;
+                };
+            };
+        } else {
+            [toLowerANSI _subMunit] call _configSearchFunc;
+        };
+    };
+    private _allMagazineConfigs = configProperties [configFile >> "cfgMagazines", "isClass _x", true];
 
-        _ammoConfig = configFile >> "CfgAmmo" >> _ammo;
+    {
+        [toLowerANSI getText (_x >> "ammo")] call _configSearchFunc;
```

Comment:
If there are only ANSI chars, then non-ANSI is the wrong tool.

https://community.bistudio.com/wiki/Code_Optimisation#toLower.2FtoUpper_vs_toLowerANSI.2FtoUpperANSI shows the ANSI versions are much faster with a yellow light, and https://community.bistudio.com/wiki/toLowerANSI says it could be 3x faster.

Reply:
Yeah, I just ran withit after rautamiekka recommended it

Reply:
Yeah, I just ran withit after rautamiekka recommended it, is there a reason not to?

Reply:
Ok, upon review, it does more than just a binary OR to set the characters to upper case. It doesn't seem to have issues with special characters, simplified Chinese, or what I think is Hiragana, it does struggle with cyrillic, so it might be better to role it back if that would be an issue.
![image](https://github.com/acemod/ACE3/assets/11843718/4484497a-d266-42cf-962b-8f8bdbf0f451)

Reply:
I know what the performance benefits are from the `toLowerANSI` vs `toLower`, you missed the point I was trying to make.
Why are you making everything lowercase?

Reply:
In my defence it was 50-50 what you coulda meant, I still think it was more like the former.

Reply:
Ignoring other problems, the toLower is to cover for cases when submunitionAmmo entires and the ammo classnames don't have the same case, thus getting added to the debug list twice.

I do see that this is covered with another check (lines 65/69), but I'd rather do it properly once and then trust it then do it once and then check a second time. So I plan on removing that.

Reply:
> when submunitionAmmo entires and the ammo classnames don't have the same case
Convert the submunitionAmmo entires to config case.

Reply:
You want me to go through every mod out there and do that?

Reply:
I could construct the config class and call `configName` on each, but that seems like more of an overhead than toLowerAnsi.

Reply:
Done